File: src/content/docs/topics/cs/coding-problems/sliding-window/003-longest-substring-without-repeating-characters.go
package main

import "fmt"

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func lengthOfLongestSubstring(s string) int {
	// TODO: implement
	return 0
}

func runTests() {
	assert(lengthOfLongestSubstring("abcabcbb") == 3)
	assert(lengthOfLongestSubstring("bbbbb") == 1)
	assert(lengthOfLongestSubstring("pwwkew") == 3)
	assert(lengthOfLongestSubstring("") == 0)
	assert(lengthOfLongestSubstring("a") == 1)
	assert(lengthOfLongestSubstring("abcdef") == 6)
	fmt.Println("all tests pass")
}

func main() {
	runTests()
}

File: src/content/docs/topics/cs/coding-problems/sliding-window/003-longest-substring-without-repeating-characters.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function lengthOfLongestSubstring(s: string): number {
    // TODO: implement
    return 0;
}

function _runTests(): void {
    assert(lengthOfLongestSubstring('abcabcbb') === 3);
    assert(lengthOfLongestSubstring('bbbbb') === 1);
    assert(lengthOfLongestSubstring('pwwkew') === 3);
    assert(lengthOfLongestSubstring('') === 0);
    assert(lengthOfLongestSubstring('a') === 1);
    assert(lengthOfLongestSubstring('abcdef') === 6);
    // perf
    const t0 = performance.now();
    lengthOfLongestSubstring('abcd'.repeat(250_000));
    console.log(`perf 1000000-char string: ${(performance.now() - t0).toFixed(1)}ms`);
    console.log('all tests pass');
}

_runTests();
